File: app/src/main/java/com/ducluanxutrieu/quanlynhanvien/Adapter/FriendsListAdapter.java
package com.ducluanxutrieu.quanlynhanvien.Adapter;

import android.app.ActivityOptions;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.ducluanxutrieu.quanlynhanvien.Activity.ChatsActivity;
import com.ducluanxutrieu.quanlynhanvien.Models.Friend;
import com.ducluanxutrieu.quanlynhanvien.R;
import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.firebase.ui.database.FirebaseRecyclerOptions;

public class FriendsListAdapter extends FirebaseRecyclerAdapter<Friend, FriendsListAdapter.ItemViewHolder> {

    public FriendsListAdapter(@NonNull FirebaseRecyclerOptions<Friend> options) {
        super(options);
    }


    @NonNull
    @Override
    public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
        View view = inflater.inflate(R.layout.item_friend, viewGroup, false);

        return new ItemViewHolder(view);
    }

    @Override
    protected void onBindViewHolder(@NonNull ItemViewHolder holder, int position, @NonNull final Friend model) {
        holder.setNameMessage(model.getName());
        holder.setRecentMessage(model.getRecentMessage());
        holder.setAvatar();
        Glide.with(holder.itemView.getContext()).load(model.getAvatarUrl()).apply(RequestOptions.circleCropTransform()).into(holder.avatar);

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Context context = v.getContext();
                Intent intent = new Intent(context, ChatsActivity.class);
                intent.putExtra("friend", model);
                ActivityOptions options =
                        ActivityOptions.makeCustomAnimation(context, R.anim.fui_slide_in_right, R.anim.fui_slide_out_left);
                context.startActivity(intent, options.toBundle());
            }
        });
    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
        TextView nameMessage;
        TextView recentMessage;
        ImageView avatar;
        View view;
        ItemViewHolder(@NonNull View itemView) {
            super(itemView);
            view = itemView;
        }

        void setNameMessage(String name) {
            nameMessage = view.findViewById(R.id.friend_name);
            nameMessage.setText(name);
        }

        void setRecentMessage(String recent) {
            recentMessage = view.findViewById(R.id.recent_chat);
            recentMessage.setText(recent);
        }

        void setAvatar() {
            avatar = view.findViewById(R.id.avatar);
        }
    }
}

File: app/src/main/java/com/ducluanxutrieu/quanlynhanvien/adapter/FriendsListAdapter.kt
package com.ducluanxutrieu.quanlynhanvien.adapter

import android.app.ActivityOptions
import android.content.Context
import android.content.Intent
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.ducluanxutrieu.quanlynhanvien.activity.ChatsActivity
import com.ducluanxutrieu.quanlynhanvien.models.Friend
import com.ducluanxutrieu.quanlynhanvien.R
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.firebase.ui.database.FirebaseRecyclerOptions
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase

class FriendsListAdapter(options: FirebaseRecyclerOptions<Friend>) : FirebaseRecyclerAdapter<Friend, FriendsListAdapter.ItemViewHolder>(options) {
    private var rootView: View? = null

    val context: Context
        get() {
            while (rootView!!.context == null) {
                try {
                    Thread.sleep(100)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }

            }
            return rootView!!.context
        }


    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ItemViewHolder {
        val inflater = LayoutInflater.from(viewGroup.context)
        rootView = inflater.inflate(R.layout.item_friend, viewGroup, false)

        return ItemViewHolder(rootView!!)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int, model: Friend) {
        holder.setNameMessage(model.name)
        holder.setRecentMessage(model.recentMessage)
        holder.setAvatar()
        Glide.with(holder.itemView.context).load(model.avatarUrl).apply(RequestOptions.circleCropTransform()).into(holder.avatar)

        holder.itemView.setOnClickListener { v ->
            val context = v.context
            val intent = Intent(context, ChatsActivity::class.java)
            intent.putExtra("friend", model)
            intent.putExtra("uid", model.uid)
            intent.putExtra("name", model.name)
            val options = ActivityOptions.makeCustomAnimation(context, R.anim.fui_slide_in_right, R.anim.fui_slide_out_left)
            context.startActivity(intent, options.toBundle())
        }
    }

    fun restoreItem(friend: Friend, key: String) {
        val reference = FirebaseDatabase.getInstance().reference
        reference.child("friend_ship/" + FirebaseAuth.getInstance().uid + "/" + key).setValue(friend)
    }

    inner class ItemViewHolder(var view: View) : RecyclerView.ViewHolder(view) {
        private lateinit var nameMessage: TextView
        private lateinit var recentMessage: TextView
        lateinit var avatar: ImageView

        fun setNameMessage(name: String?) {
            nameMessage = view.findViewById(R.id.friend_name)
            nameMessage.text = name
        }

        fun setRecentMessage(recent: String?) {
            recentMessage = view.findViewById(R.id.recent_chat)
            recentMessage.text = recent
        }

        fun setAvatar() {
            avatar = view.findViewById(R.id.avatar)
        }
    }
}
